File: patch.py
import re

with open('backend/server.js', 'r', encoding='utf-8') as f:
    content = f.read()

new_code = '''// ==========================================
// RAG MANAGER - THUẬT TOÁN CHUNKING & ROUTES
// ==========================================
function chunkTextWithOverlap(text, chunkSize = 1000, overlap = 150) {
    if (!text) return [];
    const cleanText = text.replace(/\\s+/g, ' ').trim();
    const rawSentences = cleanText.match(/[^.!?]+[.!?]+|\\s*[^.!?]+$/g) || [cleanText];
    const sentences = [];
    for (const raw of rawSentences) {
        let textToSplit = raw.trim();
        while (textToSplit.length > chunkSize) {
            let splitIndex = textToSplit.lastIndexOf(' ', chunkSize);
            if (splitIndex === -1 || splitIndex === 0) splitIndex = chunkSize;
            sentences.push(textToSplit.substring(0, splitIndex).trim());
            textToSplit = textToSplit.substring(splitIndex).trim();
        }
        if (textToSplit) sentences.push(textToSplit);
    }
    const chunks = [];
    let currentChunk = "";
    for (let i = 0; i < sentences.length; i++) {
        const sentence = sentences[i].trim();
        if (!sentence) continue;
        if (currentChunk.length + sentence.length <= chunkSize) {
            currentChunk += (currentChunk ? " " : "") + sentence;
        } else {
            if (currentChunk) chunks.push(currentChunk);
            let overlapText = "";
            let j = i - 1;
            while (j >= 0 && overlapText.length + sentences[j].length <= overlap) {
                overlapText = sentences[j].trim() + " " + overlapText;
                j--;
            }
            currentChunk = (overlapText ? overlapText.trim() + " " : "") + sentence;
        }
    }
    if (currentChunk) chunks.push(currentChunk);
    return chunks;
}

const ragController = {
    uploadAndVectorizeDocument: async (req, res) => {
        if (!req.file) return res.status(400).json({ success: false, error: "Thiếu tệp đính kèm." });
        const fileName = req.file.originalname;
        const fileSize = req.file.size;
        if (!fileName.toLowerCase().endsWith('.txt') || req.file.mimetype !== 'text/plain') {
            return res.status(400).json({ success: false, error: "Chỉ hỗ trợ định dạng .txt." });
        }
        if (fileSize > 500 * 1024) return res.status(400).json({ success: false, error: "Tệp tin vượt quá 500KB." });
        const textContent = req.file.buffer.toString('utf-8');
        if (!textContent.trim()) return res.status(400).json({ success: false, error: "Tập tin rỗng." });

        const client = await pool.connect();
        try {
            await client.query('BEGIN');
            const metadataSql = "INSERT INTO rag_documents (file_name, file_size, chunk_count, uploader_id, created_at) VALUES ($1, $2, $3, $4, NOW()) RETURNING id";
            const docResult = await client.query(metadataSql, [fileName, fileSize, 0, req.user.id]);
            const documentId = docResult.rows[0].id;
            const chunks = chunkTextWithOverlap(textContent, 1000, 150);
            if (chunks.length === 0) throw new Error("Không thể trích xuất dữ liệu.");

            const BATCH_SIZE = 20; 
            const departmentCode = req.user.department_code || 'GLOBAL';
            let successCount = 0;

            for (let i = 0; i < chunks.length; i += BATCH_SIZE) {
                const batchChunks = chunks.slice(i, i + BATCH_SIZE);
                const batchEmbeddings = await Promise.all(
                    batchChunks.map(async (chunk) => {
                         const vector = await generateEmbedding(chunk); 
                         if (!vector) throw new Error("API Nhúng Vector thất bại.");
                         return vector;
                    })
                );
                for (let j = 0; j < batchChunks.length; j++) {
                    const formatEmbedding = "[" + batchEmbeddings[j].join(',') + "]";
                    const insertSql = "INSERT INTO company_knowledge_base (document_id, content, embedding, source_type, metadata, created_at) VALUES ($1, $2, $3::vector, $4, $5, NOW())";
                    const chunkMetadata = { department_code: departmentCode, chunk_index: i + j, total_chunks: chunks.length };
                    await client.query(insertSql, [documentId, batchChunks[j], formatEmbedding, 'DOCUMENT_UPLOAD', JSON.stringify(chunkMetadata)]);
                    successCount++;
                }
                if (i + BATCH_SIZE < chunks.length) await new Promise(resolve => setTimeout(resolve, 500));
            }
            await client.query("UPDATE rag_documents SET chunk_count = $1 WHERE id = $2", [successCount, documentId]);
            await client.query('COMMIT');
            res.json({ success: true, message: "Đã nhúng thành công " + successCount + " chunks.", document_id: documentId });
        } catch (error) {
            await client.query('ROLLBACK');
            res.status(500).json({ success: false, error: "Dịch vụ AI gián đoạn." });
        } finally {
            client.release();
        }
    },
    getDocuments: async (req, res) => {
        try {
            const sql = "SELECT id, file_name, file_size, chunk_count, uploader_id, created_at FROM rag_documents ORDER BY created_at DESC";
            const { rows } = await pool.query(sql);
            res.json({ success: true, data: rows });
        } catch (error) {
            res.status(500).json({ success: false, error: "Lỗi máy chủ khi lấy danh sách." });
        }
    },
    deleteDocument: async (req, res) => {
        try {
            const sql = "DELETE FROM rag_documents WHERE id = $1 RETURNING id";
            const { rows } = await pool.query(sql, [req.params.id]);
            if (rows.length === 0) return res.status(404).json({ success: false, error: "Dữ liệu không tồn tại." });
            res.json({ success: true, message: "Xóa thành công." });
        } catch (error) {
            res.status(500).json({ success: false, error: "Lỗi máy chủ khi xóa." });
        }
    }
};

app.post('/api/rag/upload', authenticateUser, checkAdmin, upload.single('file'), ragController.uploadAndVectorizeDocument);
app.get('/api/rag/documents', authenticateUser, checkAdmin, ragController.getDocuments);
app.delete('/api/rag/documents/:id', authenticateUser, checkAdmin, ragController.deleteDocument);
'''

start_str = "app.post('/api/rag/upload'"
end_str = "app.post('/api/ai/auto-tasking'"

start_idx = content.find(start_str)
if start_idx == -1:
    print("Could not find start string")
    exit(1)

comment_idx = content.rfind("//", 0, start_idx)
if comment_idx != -1 and (start_idx - comment_idx) < 100:
    start_idx = comment_idx

end_idx = content.find(end_str, start_idx)
if end_idx == -1:
    print("Could not find end string")
    exit(1)

content = content[:start_idx] + new_code + "\n\n" + content[end_idx:]

with open('backend/server.js', 'w', encoding='utf-8') as f:
    f.write(content)

print("Successfully patched server.js")

File: patch.js
const fs = require('fs');
const path = require('path');
const controllerPath = path.join('backend', 'src', 'controllers', 'taskController.js');
let code = fs.readFileSync(controllerPath, 'utf8');

if (!code.includes('const pool = require')) {
  code = "const pool = require('../config/db');\n" + code;
}

const newHandlers = `
const updateTaskStatusHandler = async (req, res) => {
  try {
    const { id } = req.params;
    const { status, evidence } = req.body;
    const updateQuery = 'UPDATE tasks SET status = $1, updated_at = NOW() WHERE id = $2 RETURNING *';
    const { rows } = await pool.query(updateQuery, [status, id]);
    res.json({ success: true, data: rows[0] });
  } catch (error) {
    res.status(500).json({ error: 'Lỗi server khi cập nhật trạng thái.' });
  }
};

const deleteTaskHandler = async (req, res) => {
  try {
    const { id } = req.params;
    if (!req.user || req.user.role !== 'SUPER_ADMIN') {
      return res.status(403).json({ success: false, error: 'Chỉ SUPER_ADMIN mới có quyền xóa vĩnh viễn công việc.' });
    }
    await pool.query('DELETE FROM task_comments WHERE task_id = $1', [id]);
    const { rowCount } = await pool.query('DELETE FROM tasks WHERE id = $1', [id]);
    if (rowCount === 0) return res.status(404).json({ success: false, error: 'Không tìm thấy công việc.' });
    res.json({ success: true, message: 'Đã xóa công việc vĩnh viễn.' });
  } catch (error) {
    res.status(500).json({ success: false, error: 'Lỗi server khi xóa công việc.' });
  }
};

const updateTaskSupportHandler = async (req, res) => {
  try {
    const { id } = req.params;
    const updateQuery = 'UPDATE tasks SET needs_support = true, updated_at = NOW() WHERE id = $1 RETURNING id, title, needs_support as "needsSupport"';
    const { rows } = await pool.query(updateQuery, [id]);
    res.json({ success: true, message: 'Đã gửi yêu cầu hỗ trợ', data: rows[0] });
  } catch (error) {
    res.status(500).json({ error: 'Lỗi máy chủ nội bộ' });
  }
};

const restoreTaskHandler = async (req, res) => {
  try {
    const taskId = req.params.id;
    const { deadline } = req.body;
    if (!deadline) return res.status(400).json({ success: false, error: 'Bắt buộc phải có Deadline mới để khôi phục công việc.' });
    
    const updateQuery = "UPDATE tasks SET status = 'todo', deadline = $1, completed_at = NULL, updated_at = NOW() WHERE id = $2 RETURNING id, title, status, deadline";
    const { rows: updatedRows } = await pool.query(updateQuery, [deadline, taskId]);
    
    await pool.query('INSERT INTO task_comments (task_id, user_id, content, created_at) VALUES ($1, $2, $3, NOW())', [taskId, req.user.id, \`🔄 [HỆ THỐNG]: Công việc được KHÔI PHỤC về trạng thái TODO với Deadline gia hạn tới: \${deadline}\`]);
    res.json({ success: true, data: updatedRows[0] });
  } catch (err) {
    res.status(500).json({ success: false, error: 'Lỗi máy chủ khi khôi phục công việc.' });
  }
};

const getTaskCommentsHandler = async (req, res) => {
  try {
    const { id } = req.params;
    const { rows } = await pool.query(\`
      SELECT c.*, u.full_name as user_name, r.name as user_role 
      FROM task_comments c 
      LEFT JOIN users u ON c.user_id = u.id 
      LEFT JOIN roles r ON u.role_id = r.id
      WHERE c.task_id = $1 
      ORDER BY c.created_at ASC
    \`, [id]);
    res.json({ success: true, data: rows });
  } catch (err) {
    res.status(500).json({ success: false, error: 'Lỗi tải bình luận: ' + err.message });
  }
};

const addTaskCommentHandler = async (req, res) => {
  try {
    const { id } = req.params;
    const comment = req.body.comment || req.body.content;
    if (!comment) return res.status(400).json({ error: 'Nội dung bình luận trống' });
    
    const realUserId = req.user.id;
    const { rows } = await pool.query('INSERT INTO task_comments (task_id, user_id, content) VALUES ($1, $2, $3) RETURNING *', [id, realUserId, comment]);
    const newCommentId = rows[0]?.id;
    
    if (newCommentId) {
      const fullComment = await pool.query(\`
        SELECT c.*, u.full_name as user_name, r.name as user_role 
        FROM task_comments c 
        LEFT JOIN users u ON c.user_id = u.id 
        LEFT JOIN roles r ON u.role_id = r.id 
        WHERE c.id = $1
      \`, [newCommentId]);
      return res.json({ success: true, data: fullComment.rows[0] });
    } else {
      return res.status(500).json({ success: false, error: 'Không thể tạo bình luận' });
    }
  } catch (error) {
    res.status(500).json({ error: 'Lỗi server khi tạo bình luận.' });
  }
};
`;

code = code.replace(/module\.exports = \{.*?\};/s, newHandlers + '\nmodule.exports = { getTasksHandler, getTasksHistoryHandler, createTaskHandler, updateTaskStatusHandler, deleteTaskHandler, updateTaskSupportHandler, restoreTaskHandler, getTaskCommentsHandler, addTaskCommentHandler };');

fs.writeFileSync(controllerPath, code, 'utf8');
console.log('taskController.js updated');

const routesPath = path.join('backend', 'src', 'routes', 'taskRoutes.js');
let routesCode = fs.readFileSync(routesPath, 'utf8');

const newRoutes = `
router.put('/:id/status', authGuard, rbacGuard, taskController.updateTaskStatusHandler);
router.delete('/:id', authGuard, rbacGuard, taskController.deleteTaskHandler);
router.put('/:id/support', authGuard, rbacGuard, taskController.updateTaskSupportHandler);
router.patch('/:id/restore', authGuard, rbacGuard, taskController.restoreTaskHandler);
router.get('/:id/comments', authGuard, rbacGuard, taskController.getTaskCommentsHandler);
router.post('/:id/comments', authGuard, rbacGuard, taskController.addTaskCommentHandler);
`;

routesCode = routesCode.replace('module.exports = router;', newRoutes + '\nmodule.exports = router;');
fs.writeFileSync(routesPath, routesCode, 'utf8');
console.log('taskRoutes.js updated');
